### PR TITLE
[Form][FrameworkBundle][TwigBundle] Add Twig filter, form-type extension and improve service definitions for HtmlSanitizer

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

This PR borrows the twig filter and the form-type extension from https://github.com/tgalopin/html-sanitizer-bundle/

Given the form builder (the input), it allows doing:
```php
$builder
    ->add('content', TextareaType::class, ['sanitize_html' => true])
;
```

And on the template side (the output), it allows doing:
```jinja
<div>
    {{ html|sanitize_html }}
</div>
```

In order to be able to wire the corresponding services, I had to change the way html-sanitizer is wired by framework-bundle:
What we need here is a default name that we can rely on. By making the default name configurable, the current way to configure html_sanitizer makes it hard for other bundles (twig-bundle here) to know what the default name is. Eg here I would have to make the bundle read the config settings of framework-bundle. I solved this issue by making the default name non-configurable - aka I'm creating a convention: the default name is "default", and this "default" sanitizer is the one aliased to the html_sanitizer service, or to the non-named autowiring alias.

I'm submitting this PR to 6.1 because html-sanitizer is new. If we prefer merging this in 6.2, the updated service wiring must be merged into 6.1 at least.

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    public static function getExtendedTypes(): iterable
+    {
+        return [TextType::class];
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver
+            ->setDefaults(['sanitize_html' => false, 'sanitizer' => 'default'])
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php]: `'sanitizer' => $this->defaultName` ?

Comment (nicolas-grekas) [reply src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php]: fixed thanks!

**Review by yceruto — APPROVED** [review]

**Review by tgalopin — APPROVED** [review]
Thanks!

(small comment: my email is galopintitouan@gmail.com now 😊)

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+use Twig\TwigFilter;
+
+/**
+ * @author Titouan Galopin <tgalopin@gmail.com>
+ */
+final class HtmlSanitizerExtension extends AbstractExtension
+{
+    public function __construct(
+        private ContainerInterface $sanitizers,
+        string $defaultName = 'default',
```
Comment (Kocal) [inline src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]:
```suggestion
        private string $defaultName = 'default',
```

No?

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    ) {
+    }
+
+    public function getFilters(): array
+    {
+        return [
+            new TwigFilter('sanitize_html', $this->sanitize(...), ['is_safe' => ['html']]),
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php:32]: IIRC, Twig does not yet support the first-class callable notation to compile extension methods to the optimized code that does not require loading all filters at runtime.

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: why making the default name configurable here if you remove it from the FrameworkBundle configuration ?

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php:32]: We use this notation for all filters in 6.1.

> compile extension methods to the optimized code

I had a look at twig's source and I'm not sure what you mean. Do you have a pointer to what you're thinking of?

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: for decoupling

Comment (stof) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: but then, why not keeping the default config configurable in FrameworkBundle ?

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: What we need here is a default name that we can rely on. By making the default name configurable, the current way to configure html_sanitizer makes it hard for other bundles (twig-bundle here) to know what the default name is. Eg here I would have to make the bundle read the config settings of framework-bundle.

I solved this issue by making the default name non-configurable - aka I'm creating a convention: the default name is "default", and this "default" sanitizer is the one aliased to the `html_sanitizer` service, or to the non-named autowiring alias.

Does that make sense to you? Would you have a better idea?

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php:32]: See `Twig\Node\Expression\CallExpression::compileCallable`

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php (hunk 2 above)
Comment (stof) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: fine with me

Comment (stof) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: this is indeed needed, to avoid using an undefined property

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php]: fixed

File: src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Twig/Extension/HtmlSanitizerExtension.php:32]: Thanks, should be fixed by https://github.com/twigphp/Twig/pull/3702

**Comment by nicolas-grekas** [comment]
PR is now tested and ready, failures unrelated.

Status: needs review

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        if (!$options['sanitize_html']) {
+            return;
+        }
+
+        $sanitizers = $this->sanitizers;
+        $sanitizer = $options['sanitizer'] ?? $this->defaultSanitizer;
```
Comment (wouterj) [inline src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php:54]: Is there a specific reason not to configure `$this->defaultSanitizer` as the default value in `configureOptions()`?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/html_sanitizer.php [hunk 5]
```diff
@@ -13,13 +13,18 @@
...
 return static function (ContainerConfigurator $container) {
     $container->services()
-        ->set('html_sanitizer.config', HtmlSanitizerConfig::class)
+        ->set('html_sanitizer.config.default', HtmlSanitizerConfig::class)
             ->call('allowSafeElements')
 
-        ->set('html_sanitizer', HtmlSanitizer::class)
+        ->set('html_sanitizer.sanitizer.default', HtmlSanitizer::class)
             ->args([service('html_sanitizer.config')])
+            ->tag('html_sanitizer', ['name' => 'default'])
```
Comment (michaljusiega) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/html_sanitizer.php:25]: After upgrade to 6.1.0-RC1 I got:
obraz

Looks like the args for HtmlSanitizer service may be invalid and IMO should be:

```suggestion
        ->set('html_sanitizer.sanitizer.default', HtmlSanitizer::class)
            ->args([service('html_sanitizer.config.default')])
            ->tag('html_sanitizer', ['name' => 'default'])
```

and it's works for me.

File: src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Form/Extension/HtmlSanitizer/Type/TextTypeHtmlSanitizerExtension.php:54]: Yes: to make it clear that `null` means "use the default value". I had removed the possibility to pass null before, and that made it impossible to say so.

**Pull request merged**